File: cine/SergioDeIscar-cine-java/SergioDeIscar-cine-java/src/main/java/models/Sala.java
package models;

import models.*;
import enums.*;
import javafx.util.Pair;

public class Sala {
    private String name;
    private Film film;
    private final Butaca[][] butacas;

    //region Constructor
    public Sala(String name, Film film, Pair<Integer,Integer> size){
        setName(name);
        setFilm(film);
        butacas = new Butaca[size.getKey()][size.getValue()];
        for (int i = 0; i < size.getKey(); i++) {
            for (int j = 0; j < size.getValue(); j++) {
                butacas[i][j] = new Butaca();
            }
        }
    }
    public Sala(String name, Film film, Pair<Integer,Integer> size, int probability){
        setName(name);
        setFilm(film);
        butacas = new Butaca[size.getKey()][size.getValue()];
        for (int i = 0; i < size.getKey(); i++) {
            for (int j = 0; j < size.getValue(); j++) {
                butacas[i][j] = new Butaca(probability);
            }
        }
    }
    //endregion

    //region Getter
    public String getName(){
        return this.name;
    }
    public Film getFilm(){
        return this.film;
    }
    public Butaca[][] getButacas(){
        return this.butacas;
    }
    //endregion

    //region Setter
    public void setName(String newName){
        this.name = newName;
    }
    public void setFilm (Film newFilm){
        this.film = newFilm;
    }
    //endregion

    //region Methods
    //region Output
    public void mostrarSala(){
        //General Info
        System.out.println("Disposición en la sala " + getName() + ":");
        System.out.println("Para la película " + getFilm().getTitulo() + ":");

        //Mapeado de la sala
        MostrarButacas();

        System.out.println("Esto da un total de:\n" + countEstadosText());
        System.out.println("Y un balance total de: " + getBalance() + "€");
    }

    private void MostrarButacas() {
        for (int i = 0; i < butacas.length - 1; i++) {
            for (int j = 0; j < butacas[0].length; j++) {
                char vipMarc = butacas[i][j].getIsVip() ? '*' : ' ';
                switch (butacas[i][j].getEstado()){
                    case LIBRE -> System.out.print("\t| " + vipMarc + " L " + vipMarc + " |\t");
                    case RESERVADA -> System.out.print("\t| " + vipMarc + " R " + vipMarc + " |\t");
                    case OCUPADA -> System.out.print("\t| " + vipMarc + " O " + vipMarc + " |\t");
                }
            }
            System.out.println();
        }
    }
    //endregion

    //region CountEstados
    public String countEstadosText(){
        int[] countEstados = countEstados();
        return  "\t" + countEstados[0] + " butacas libres.\n" +
                "\t" + countEstados[1] + " butacas reservadas.\n" +
                "\t" + countEstados[2] + " butacas ocupadas.";
    }

    public int[] countEstados(){
        int[] newArray = new int[3];
        newArray[0] = countButacas(ButacaEstado.LIBRE);
        newArray[1] = countButacas(ButacaEstado.RESERVADA);
        newArray[2] = countButacas(ButacaEstado.OCUPADA);
        return newArray;
    }

    private int countButacas(ButacaEstado estado) {
        int count = 0;
        for (Butaca[] i : butacas){
            for (Butaca j : i){
                if (j.getEstado() == estado) count++;
            }
        }
        return count;
    }
    //endregion

    //region getBalance
    public Float getBalance(){
        return getBalance(5.35f, 8.5f);
    }
    public Float getBalance(Float normalPrice, Float vipPrice){
        float count = 0f;
        for (Butaca[] i : butacas){
            for (Butaca j : i){
                if (j.getEstado() == ButacaEstado.OCUPADA) count += j.getIsVip() ? vipPrice : normalPrice;
            }
        }
        return count;
    }
    //endregion

    //region CRUD Butaca
    private boolean editButaca(Pair<Integer, Integer> pos, ButacaEstado estadoEqual, ButacaEstado newEstado){
        if (butacas[pos.getValue()][pos.getKey()] == null) return false;
        if(pos.getValue() > butacas.length - 1 || pos.getKey() > butacas[0].length - 1) return false;
        if (butacas[pos.getValue()][pos.getKey()].getEstado() != estadoEqual) return false;
        butacas[pos.getValue()][pos.getKey()].setEstado(newEstado);
        return true;
    }

    public boolean reservaButaca(Pair<Integer, Integer> pos){
        return editButaca(pos, ButacaEstado.LIBRE, ButacaEstado.RESERVADA);
    }

    public boolean formalizarReserva(Pair<Integer, Integer> pos){
        return editButaca(pos, ButacaEstado.RESERVADA, ButacaEstado.OCUPADA);
    }

    public boolean anularReserva(Pair<Integer, Integer> pos){
        if (butacas[pos.getValue()][pos.getKey()] == null) return false;
        if(pos.getValue() > butacas.length - 1 || pos.getKey() > butacas[0].length - 1) return false;
        if (butacas[pos.getValue()][pos.getKey()].getEstado() == ButacaEstado.LIBRE) return false;
        butacas[pos.getValue()][pos.getKey()].setEstado(ButacaEstado.LIBRE);
        return true;
    }
    //endregion

    /**
     * Hace la traducción de número a carácter según el alfabeto español.
     * 0 -> A; 26 -> Z; AA -> 27
     */
    public String alphabetNumberToString(int num) {
        String alphabet = "ABCDEFGHIJKLMNÑOPQRSTUVWXYZ";
        StringBuilder result = new StringBuilder();
        var number = num;
        while (number >= 0) {
            result.append(alphabet.indexOf(number % alphabet.length()));
            number /= alphabet.length();
            number -= 1;
        }
        return result.reverse().toString();
    }

    public Pair<Integer, Integer> getSize(){
        return new Pair<>(butacas[0].length - 1, butacas.length - 1);
    }
    //endregion
}

File: cine/SergioDeIscar-cine-kotlin/SergioDeIscar-cine-kotlin/src/main/kotlin/salacine/models/Sala.kt
package salacine.models

import salacine.enums.ButacaEstado

data class Sala(val nombre: String, val film: Film, val size: Pair<Int, Int>){

    private var butacas: Array<Array<Butaca>> = Array(size.second){ Array(size.first){ Butaca(ButacaEstado.LIBRE) } }

    /**
     * Constructor secundario para indicar la probabilidad de que una butaca sea vip
     * @param size Par que indica la X -> fila y la Y -> columnas que va a tener la sala
     * @param vipProbability Rango que marca la probabilidad de que una butaca sea vip, la escala va de 0-100
     */
    constructor(nombre: String, film: Film, size: Pair<Int, Int>, vipProbability: Int): this(nombre, film, size) {
        butacas = Array(this.size.second){ Array(this.size.first){ Butaca(ButacaEstado.LIBRE, vipProbability) } }
    }

    fun mostrarSala(){
        // region Información general
        println("Disposición en la sala ${this.nombre}:")
        println("Para la película ${film.titulo}:\n")
        // endregion

        mostrarButacas()

        // region Información en cuanto a números
        println("Esto da un total de:\n" + countEstadosText())
        println("Y un balance total de ${balanceSala()} €.")
        // endregion
    }

    fun mostrarButacas() {
        mostrarIndices()
        println()
        for (i in butacas.indices.last downTo 0) {
            print("\t|  ${alphabetNumberToString(i)}  |\t")
            for (j in butacas[i].indices) {

                val vipMarc = if (butacas[i][j].getIsVip()) '*' else ' '

                print(
                    when (butacas[i][j].getEstado()) {
                        ButacaEstado.RESERVADA -> "\t| $vipMarc R $vipMarc |\t"
                        ButacaEstado.OCUPADA -> "\t| $vipMarc O $vipMarc |\t"
                        else -> "\t| $vipMarc L $vipMarc |\t"
                    }
                )
            }
            println()
        }
    }

    private fun mostrarIndices() {
        print("\t|     |\t")
        for (i in butacas[0].indices) {
            print("\t|   ${i + 1}  |\t")
        }
    }

    fun countEstadosText(): String{
        val countEstados = countEstados()
        return  "\t${countEstados[0]} butacas libres.\n" +
                "\t${countEstados[1]} butacas reservadas.\n" +
                "\t${countEstados[2]} butacas ocupadas."
    }

    /**
     * Hace la traducción de número a carácter según el alfabeto español.
     * 0 -> A; 26 -> Z; AA -> 27
     */
    fun alphabetNumberToString(num: Int): String {
        val alphabet = "ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"
        val result = StringBuilder()
        var number = num
        while (number >= 0) {
            result.append(alphabet[number % alphabet.length])
            number /= alphabet.length
            number -= 1
        }
        return result.reverse().toString()
    }

    /**
     * Función que edita el estado de una determinada butaca.
     * @param pos Es un par que almacena la X y la Y de la butaca, es decir su posición
     * @param estadoEqual Determina que actualmente la butaca requiere estar en ese estado
     * @param newEstado Es el nuevo estado de la butaca
     */
    private fun editButaca(pos: Pair<Int, Int>, estadoEqual: ButacaEstado, newEstado: ButacaEstado): Boolean{
        if(pos.second > butacas.lastIndex || pos.first > butacas[0].lastIndex) return false
        if (butacas[pos.second][pos.first].getEstado() != estadoEqual) return false
        butacas[pos.second][pos.first].setEstado(newEstado)
        return true
    }

    fun reservaButaca(pos: Pair<Int, Int>): Boolean{
        return editButaca(pos, ButacaEstado.LIBRE, ButacaEstado.RESERVADA)
    }

    fun formalizarReserva(pos: Pair<Int, Int>): Boolean{
        return editButaca(pos, ButacaEstado.RESERVADA, ButacaEstado.OCUPADA)
    }

    fun anularReserva(pos: Pair<Int, Int>): Boolean{
        if(pos.second > butacas.lastIndex || pos.first > butacas[0].lastIndex) return false
        if (butacas[pos.second][pos.first].getEstado() == ButacaEstado.LIBRE) return false
        butacas[pos.second][pos.first].setEstado(ButacaEstado.LIBRE)
        return true
    }

    internal fun countEstados(): IntArray{
        val newArray = IntArray(3)
        newArray[0] = countButacas(ButacaEstado.LIBRE)
        newArray[1] = countButacas(ButacaEstado.RESERVADA)
        newArray[2] = countButacas(ButacaEstado.OCUPADA)
        return newArray
    }

    fun balanceSala(precioNormal: Float = 5.35f, precioVip: Float = 8.5f): Float{
        var count = 0f
        for (i in butacas){
            for (j in i){
                if (j.getEstado() == ButacaEstado.OCUPADA) count += if (j.getIsVip()) precioVip else precioNormal
            }
        }
        return count
    }

    private fun countButacas(estado: ButacaEstado): Int {
        var count = 0
        for (i in butacas){
            for (j in i){
                if (j.getEstado() == estado) count++
            }
        }
        return count
    }

    internal fun getSize(): Pair<Int, Int> {
        return Pair(butacas[0].size, butacas.size)
    }
}